### PR TITLE
feat: VAST Data storage backend

### PR DESCRIPTION
**Pull request by fregataa**
**Checklist:** (if applicable)

- [x] Milestone metadata specifying the target backport version

### CODE REVIEW

File: src/ai/backend/storage/vast/vastdata_client.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+        current_dt = datetime.now()
+        if self._auth_token is None:
+            return await self._login()
+        elif self._auth_token["access_token"]["issued_dt"] + DEFAULT_ACCESS_TOKEN_SPAN > current_dt:
+            return
+        elif (
+            self._auth_token["refresh_token"]["issued_dt"] + DEFAULT_REFRESH_TOKEN_SPAN > current_dt
+        ):
+            return await self._refresh()
+        return await self._login()
```
Comment (kyujin-cho) [inline src/ai/backend/storage/vast/vastdata_client.py:181]: VAST's API token is formed with JWT format so we can just evaluate `exp` value of JWT payload to determine if token is valid or not.

File: src/ai/backend/storage/context.py [hunk 2]
```diff
@@ -39,6 +39,7 @@
 )
 from .purestorage import FlashBladeVolume
 from .types import VolumeInfo
+from .vast import VastVolume
```
Comment (kyujin-cho) [inline src/ai/backend/storage/context.py]: We need to call them `VAST`, not `Vast` ([link to their official webpage](https://vastdata.com/)). Capital letters do matter!

File: src/ai/backend/storage/vast/__init__.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            api_version=self.config["vast_api_version"],
+            ssl=ssl_verify,
+        )
+
+    async def create_quota_model(self) -> VastQuotaModel:
+        return VastQuotaModel(self.mount_path, self.api_client)
+
+    async def get_capabilities(self) -> FrozenSet[str]:
+        return frozenset(
+            [CAP_VFOLDER, CAP_METRIC, CAP_QUOTA, CAP_FAST_FS_SIZE, CAP_FAST_SCAN, CAP_FAST_SIZE]
```
Comment (kyujin-cho) [inline src/ai/backend/storage/vast/__init__.py]: `CAP_FAST_SCAN` capability should only be advertised if the storage supports scanning number of files in a "fast" fashion. In here fast means we should not have to scan the whole filesystem and count the number of files one by one, usually by querying the API vender has exposed. But here I don't see any of such custom methods overridden. Purestorage backend's `RapidFileToolsFSOpModel.scan_tree_usage()` can be a good example of such feature.

File: src/ai/backend/storage/vast/vastdata_client.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+                f"quotas/{vast_quota_id}/",
+                body,
+            )
+            data: Mapping[str, Any] = await response.json()
+            match response.status // 100:
+                case 2:
+                    pass
+                case 4 | 5:
+                    err_msg = data.get("detail", "Unkown error from vast API")
+                    raise VastInvalidParameterError(err_msg)
```
Comment (kyujin-cho) [inline src/ai/backend/storage/vast/vastdata_client.py]: In theory `Invalid Parameter` error should only be represented by error code 400, not whole 4xx and 5xx status codes. We should discuss whether this kind of approach is legitimate or not.

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/storage/vast/vastdata_client.py (hunk 4 above)
Comment (fregataa) [reply src/ai/backend/storage/vast/vastdata_client.py]: I updated this part as handling 4xx and 5xx in the same way would be buggy.
The reason why I handled them in the same way is that some invalid API call with wrong parameter or invalid condition among parameters (such as, setting soft limit bigger than hard limit) would responses with status code 500 but it is client's fault.

**Review by kyujin-cho — APPROVED** [review]

**Pull request merged**